### ISSUE TITLE
Add internal metric for dropped UDP packets

### ISSUE THREAD

**Issue by jszwedko** [vectordotdev/vector#15586]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

It would be useful for Vector users to know if UDP packets destined for Vector are being dropped to monitor the level of dropped data that doesn't reach Vector.

### Proposal

Publish a new internal metric for dropped UDP packets, `component_dropped_udp_packets_total`, that is tagged with the component id, type, and kind (as usual).

### References

* https://github.com/vectordotdev/vector/issues/15583

### Version

vector 0.26.0

**Comment by zamazan4ik** [comment]
At which point UDP packets are dropped? I quickly researched the `socket` source and cannot find the place where such a metric could be introduced. As far as I understand the original problem, some internal OS socket buffer exceeds some size and newer packets are dropped silently even before Vector has a chance to process them. Correct me if I am wrong :)

**Comment by jszwedko** [comment]
> At which point UDP packets are dropped? I quickly researched the `socket` source and cannot find the place where such a metric could be introduced. As far as I understand the original problem, some internal OS socket buffer exceeds some size and newer packets are dropped silently even before Vector has a chance to process them. Correct me if I am wrong :)

That is correct, the packets are dropped before Vector sees them. I'm realizing I forgot to include more context here, but Vector could determine the number of dropped packets by reading `/proc`. I'll update the description.

**Comment by zamazan4ik** [comment]
So if I am reading this edit right, it is a Linux-specific metric. Do we already have any OS-specific metrics? I guess at least we could start with Linux-only solution.

By the way, is there a way to implement it on macOS (probably yes)/Windows (not so sure) and other OSes (*BSD, whatsoever else)?

**Comment by spencergilbert** [comment]
> So if I am reading this edit right, it is a Linux-specific metric. Do we already have any OS-specific metrics? I guess at least we could start with Linux-only solution.
> 
> By the way, is there a way to implement it on macOS (probably yes)/Windows (not so sure) and other OSes (*BSD, whatsoever else)?

I don't remember which off the top of my head, but I remember seeing at least one config flag in the `host_metrics` source related to the OS, so there should be some precedence.
